lib/src/cancelable_operation.dart

```diff
@@ -262,16 +263,21 @@ class CancelableOperation<T> {
                       error2, identical(error, error2) ? stack : stack2);
                 }
               });
-    _completer._cancelCompleter?.future.whenComplete(onCancel == null
-        ? completer._cancel
+    final forwardingCancel = onCancel == null
+        ? () => !completer.isCanceled ? completer._cancel() : Future.value()
         : () async {
             if (completer.isCanceled) return;
             try {
               await onCancel(completer);
             } catch (error, stack) {
               completer.completeError(error, stack);
             }
-          });
+          };
```

Comment:
Instead of creating functions here, storing them in `_extraOnCancel`, and calling them at the other end, 
could you just store pairs of `(CancelableCompleter completer, FutureOr<void> Function(CancelableCompleter<R>)? onCancel)`, and then handle all this logic at the consumption point?
(Like, have a `Future<void>? _callExtraOnCancel(CancelableCompleter completer, FutureOr<void> Function(CancelableCompleter<R>)? onCancel) { ... }` which does what the two functions above would given the same values.)

Then the code below can be:
```dart
   if (_completer.isCanceled) {
     Timer(Duration.zero, () {
       _callExtraOnCancel(completer, onCancel);
     });
   } else {
      _completer._extraOnCancel.add((completer, onCancel));
   }
```
(May need a helper class if not 3.0 ready yet.)
It seems those two values are all the closures use anyway.

Reply:
> could you just store pairs of `(CancelableCompleter completer, FutureOr<void> Function(CancelableCompleter<R>)? onCancel)`

Do we have any syntax to express a generic record type?

I don't think this will work as a list of records without being able to match the generics between the completer and onCancel

Reply:
I added a `_CancelForwarder` class to avoid some of these eager closures.

Reply:
No generic record types. I'd probably just go with `(CancelableCompleter<Object?> completer, FutureOr<void> Function(CancelableCompleter<Object?>)?`, the value type `R` isn't needed when all we do is cancel.